contracts/nois-airdrop/src/contract.rs

```diff
@@ -0,0 +1,790 @@
+        funds: info.funds, // Just pass on all funds we got
+    });
+    Ok(response)
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    callback: NoisCallback,
+) -> Result<Response, ContractError> {
+    let RandomnessParams {
+        nois_proxy,
+        nois_randomness: _,
+        requested,
+    } = NOIS_RANDOMNESS.load(deps.storage).unwrap();
+
+    // callback should only be allowed to be called by the proxy contract
+    // otherwise anyone can cut the randomness workflow and cheat the randomness by sending the randomness directly to this contract
+    ensure_eq!(info.sender, nois_proxy, ContractError::UnauthorizedReceive);
+    let randomness: [u8; 32] = callback
+        .randomness
+        .to_array()
+        .map_err(|_| ContractError::InvalidRandomness)?;
+    // Make sure the randomness does not exist yet
+
+    match NOIS_RANDOMNESS.load(deps.storage).unwrap().nois_randomness {
+        None => NOIS_RANDOMNESS.save(
+            deps.storage,
+            &RandomnessParams {
+                nois_proxy,
+                nois_randomness: Some(randomness),
+                requested,
+            },
+        ),
+        Some(_randomness) => return Err(ContractError::ImmutableRandomness),
+    }?;
+
+    Ok(Response::default())
+}
+
+fn execute_withdraw_all(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    // check the calling address is the authorised addr
+
+    // if manager set validate, otherwise unauthorized
+    let manager = config.manager.ok_or(ContractError::Unauthorized {})?;
+    if info.sender != manager {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    let amount = deps
+        .querier
+        .query_balance(env.contract.address, AIRDROP_DENOM)?;
+    let msg = BankMsg::Send {
+        to_address: config.withdrawal_address.into(),
+        amount: vec![amount],
+    };
+    let res = Response::new()
+        .add_message(msg)
+        .add_attribute("action", "withdraw_all");
+    Ok(res)
+}
+
+fn is_randomly_eligible(addr: &Addr, randomness: [u8; 32]) -> bool {
+    let mut hasher = Sha256::new();
+    hasher.update(addr.as_bytes());
+    let sender_hash: [u8; 32] = hasher.finalize().into();
+    // Concatenate the randomness and sender hash values
+    let combined = [randomness, sender_hash].concat();
+    // Hash the combined value using SHA256 to generate a random number between 1 and 3
+    let mut hasher = Sha256::new();
+    hasher.update(&combined);
+    let hash = hasher.finalize();
+    let outcome = hash[0] % AIRDROP_ODDS;
+
+    // returns true if the address is eligible
+    matches!(outcome, 0)
+}
+
+fn execute_register_merkle_root(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    merkle_root: String,
```

Comment:
You can use the HexBinary type to make your life easier:

```suggestion
    merkle_root: HexBinary,
```

This is hex at JSON level and decodes to binary data automatically via serde.